qiskit_machine_learning/neural_networks/neural_network.py

```diff
@@ -0,0 +1,127 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2020, 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""A Neural Network abstract class for all (quantum) neural networks within Qiskit's
+machine learning module."""
+
+
+from abc import ABC, abstractmethod
+from typing import Tuple, Union, List, Optional, Dict
+
+import numpy as np
+
+from ..exceptions import QiskitMachineLearningError
+
+
+class NeuralNetwork(ABC):
+    """Abstract Neural Network class providing forward and backward pass and handling
+    batched inputs. This is to be implemented by other (quantum) neural networks.
+    """
+
+    def __init__(self, num_inputs: int, num_weights: int,
+                 output_shape: Union[int, Tuple[int, ...]]) -> None:
+        """Initializes the Neural Network.
+        Args:
+            num_inputs: The number of input features.
+            num_weights: The number of trainable weights.
+            output_shape: The shape of the output.
+        Raises:
+            QiskitMachineLearningError: Invalid parameter values.
+        """
+        if num_inputs < 0:
+            raise QiskitMachineLearningError('Number of inputs cannot be negative!')
```

Comment:
Its always nice to have the invalid value etc in these error messages of course.

Reply:
Added invalid values to the messages.